fendermint/vm/interpreter/src/fvm/cometbft.rs

```diff
@@ -0,0 +1,115 @@
+// Copyright 2022-2024 Protocol Labs
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use crate::fvm::state::FvmExecState;
+use crate::fvm::{FvmMessageInterpreter, PowerUpdates};
+use anyhow::Context;
+use fendermint_vm_genesis::{Power, Validator};
+use fvm_ipld_blockstore::Blockstore;
+use tendermint_rpc::Client;
+
+/// The end block update for cometbft
+pub struct EndBlockUpdate {
+    pub consensus: Option<tendermint::consensus::Params>,
+    pub validators: PowerUpdates,
+}
+
+/// Potential updates to cometbft consensus parameters. Currently only block `max_gas` needs to be
+/// updated, but in the future, more parameter could be updated.
+pub struct ConsensusBlockUpdate {
+    max_gas: Option<u64>,
+}
+
+/// Convert validator power to tendermint validator update.
+/// TODO: the import is quite strange, `Validator` and `Power` are imported from `genesis` crate,
+/// TODO: which should be from a `type` or `validator` crate.
+pub fn to_validator_updates(
+    validators: Vec<Validator<Power>>,
+) -> anyhow::Result<Vec<tendermint::validator::Update>> {
+    let mut updates = vec![];
+    for v in validators {
+        updates.push(tendermint::validator::Update {
+            pub_key: tendermint::PublicKey::try_from(v.public_key)?,
+            power: tendermint::vote::Power::try_from(v.power.0)?,
+        });
+    }
+    Ok(updates)
+}
+
+impl TryFrom<EndBlockUpdate> for tendermint::abci::response::EndBlock {
+    type Error = anyhow::Error;
+
+    fn try_from(value: EndBlockUpdate) -> Result<Self, Self::Error> {
+        let validator_updates = to_validator_updates(value.validators.0)
+            .context("failed to convert validator updates")?;
+
+        Ok(tendermint::abci::response::EndBlock {
+            validator_updates,
+            consensus_param_updates: value.consensus,
+            events: Vec::new(), // TODO: Events from epoch transitions?
+        })
+    }
+}
+
+impl<DB, TC> FvmMessageInterpreter<DB, TC>
+where
+    DB: Blockstore + Clone + 'static + Send + Sync,
+    TC: Client + Clone + Send + Sync + 'static,
+{
+    pub(crate) async fn update_cometbft_consensus_params(
+        &self,
+        state: &mut FvmExecState<DB>,
+        end_block: &mut EndBlockUpdate,
+    ) -> anyhow::Result<()> {
+        let mut updates = ConsensusBlockUpdate::empty();
+
+        state.gas_market().process_consensus_update(&mut updates);
+
+        if !updates.is_some() {
+            return Ok(());
+        }
+
+        let params = self
+            .client
+            .consensus_params(tendermint::block::Height::try_from(state.block_height())?)
+            .await?
+            .consensus_params;
+        end_block.with_consensus(updates.apply(params));
+
+        Ok(())
+    }
+}
```

Comment:
This looks rather over-engineered. What's wrong with having this code in `App::end_block`?

Reply:
Currently the cometbft `Client` is held by interpreter, so in[ App::end_block](https://github.com/consensus-shipyard/ipc/blob/f4c67489325ec6aa9ded608b2be9118c89cd55e5/fendermint/app/src/app.rs#L124), there is no access to client.

Also `state` is held in a [closure](https://github.com/consensus-shipyard/ipc/blob/f4c67489325ec6aa9ded608b2be9118c89cd55e5/fendermint/app/src/app.rs#L813), so adding to interpreter is less of a change now.

The `EndBlockUpdate` is added because `interpreter` expected `Output` associated type, currently it's [PowerUpdate](https://github.com/consensus-shipyard/ipc/blob/f4c67489325ec6aa9ded608b2be9118c89cd55e5/fendermint/vm/interpreter/src/fvm/exec.rs#L51). Having a struct is easier to pass around.

Now that you mentioned, maybe the direction to go is move all `tendermint` related operation to `App` crate (currently tendermint crates and fendermint crates are coupled), pave way for consensus agnostic implementation.